[COT] Reintroduce the orders cache

vedanshujain opened the pull request:
In #34992 we reverted the orders cache that we added in #34396. This PR scopes to introduce it back.

From Original PR:

### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Use [the recently implemented object cache](https://github.com/woocommerce/woocommerce/pull/33634) to cache orders. This includes:

1. A new `OrderCache` class (which is very simple since the bulk of the work is done by the base `ObjectCache` class).
2. A new `Orders cache` experimental feature (uses the recently introduced [features engine](https://github.com/woocommerce/woocommerce/pull/34616)).
3. A new `OrderCacheController` class to control the cache usage, including turning it off temporarily (this is done while synchronization is in progress).
4. Caching all orders that are retrieved via `WC_Order_Factory` (this covers the `wc_get_order(s)` functions, the `WC_Order_Query` class and `query` in the new data store class).
5. Updating the cached value of orders whenever they are saved, but only if the order had been already cached.
6. Dropping orders from the cache whenever they are trashed or deleted (this is done by hooking into `woocommerce_trash_order` and `woocommerce_delete_order`).
7. Flushing the entire cache whenever the custom orders table usage is enabled or disabled.

A check for the orders cache feature being enabled and not being temporarily disabled is made before any caching operation.

The `OrderCacheController` class includes two methods, `temporarily_disable_orders_cache_usage` and `maybe_restore_orders_cache_usage`. The former stores an in-memory variable with a backup of the cache enable cache setting, the later removes the variable. The feature isn't actually modified, the temporary variable is only used by the `orders_cache_usage_is_enabled` method, which is the one used before doing any order caching operation.

As a bonus, this pull request introduces a couple of tweaks to the base `ObjectCache` class:

* A new `update_if_cached` method.
* `set` will now validate the object to cache _before_ attempting to retrieve its id. This way, if the object is invalid the error returned will be "invalid object" instead of "can't get the id".

Closes #34100.

### How to test the changes in this Pull Request:

## Note that since PR has changed significantly from revert, test instructions no longer apply. Use the updated instructions here:


Use the following commands in a wp shell to test whether orders are getting cached as expected:

```php

define( 'SAVEQUERIES', true );

$id = XX; // Some order ID that already exists.

get_num_queries(); // Note down the current number of queries before starting our test.

$order = wc_get_order( $id );

get_num_queries(); // This would have increased by 3-4 since last test.

$r_order = wc_get_order( $id );

get_num_queries(); // This should be the same as the previous test if the order was cached successfully.

\Automattic\WooCommerce\Utilities\ArrayUtil::deep_assoc_array_diff( $order->get_data(), $r_order->get_data(), false ); // This should be an empty array. 
```


#### Test that order cache is invalidated when order changes

```php
$id = XX;

$order = wc_get_order( $id );

// Change the order status.
$order->set_status( 'completed' );

// Change billing city.
$order->set_billing_city( 'New York' );

// Save the order.
$order->save();

// Get the order again.
$r_order = wc_get_order( $id );

$r_order->get_status(); // Should be 'completed'.
$r_order->get_billing_city(); // Should be 'New York'.

```


#### Test that order cache is invalidated when order is deleted

```php

$id = XX;

$order = wc_get_order( $id );

// Delete the order.
$order->delete( true );

// Get the order again.
$r_order = wc_get_order( $id ); // Should be false.

```


#### Test that order group cache is invalidate when group function are called.

```php

define( 'SAVEQUERIES', true );

$id = XX;

// Note down the current number of queries before starting our test.
get_num_queries(); 

$order = wc_get_order( $id );

// This would have increased by 3-4 since last test.
get_num_queries(); 

// Delete the order group cache.
wc_delete_shop_order_transients();

get_num_queries();

// Get the order again.
$r_order = wc_get_order( $id );

// This should have increased now since cache was invalidated.
get_num_queries(); 
```

Additionally test that order creation, updation flows etc works as expected from the UX.

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

vedanshujain committed:
Revert "Revert "[COT] Add the orders cache" (#34992)"

This reverts commit 0b5d7ab9ad4ef09fc0ef47c5deddfe2aa4f7e5c3.

barryhughes commented on plugins/woocommerce/src/Caches/OrderCacheController.php:
```diff
@@ -0,0 +1,121 @@
+
+	use AccessiblePrivateMethods;
+
+	const FEATURE_NAME = 'orders_cache';
```
We can be explicit about the visibility level:

```suggestion
	public const FEATURE_NAME = 'orders_cache';
```

vedanshujain replied:
Note that I have merged the feature flag with HPOS feature.

barryhughes commented on plugins/woocommerce/src/Caching/ObjectCache.php:
```diff
@@ -164,13 +164,13 @@ private function create_cache_key_prefix(): string {
 	 * @throws CacheException Invalid parameter, or null id was passed and get_object_id returns null too.
 	 */
-	public function set( $id = null, $object, int $expiration = self::DEFAULT_EXPIRATION ): bool {
+	public function set( $object, $id = null, int $expiration = self::DEFAULT_EXPIRATION ): bool {
```
Is flipping the order safe? This is a public method in a non-internal class (that has been available since at least [6.9.0](https://github.com/woocommerce/woocommerce/blob/6.9.0/plugins/woocommerce/src/Caching/ObjectCache.php#L173)).

vedanshujain replied:
It's probably safe because:
1. This is part of an abstract class without any actual core usage.
2. There are a few bugs that I have fixed now as part of this PR, so probably it's not getting used in the wild

That said, we can also revert if that makes more sense.

barryhughes replied:
I also flagged a 'similar thing' elsewhere, but probably you're right. Felt it was worth pausing and asking, though.

barryhughes reviewed (changes requested):
Left a couple of minor notes—but in general this looks great and works well (I also tested alongside Redis). One other question I have relates to how this feature is exposed in the admin UI:

<img width="963" alt="order-cache-feature-warning" src="https://user-images.githubusercontent.com/3594411/213002756-d2582036-1bc5-4307-9b64-f4fe5fc078be.png">

That is, right now it seems plugins would need to explicitly declare compatibility. Is that something we would want to address in this PR, or handle some other way (I realize we may anyway be revising this messaging for the HPOS feature).

vedanshujain committed:
Move the namespace algorithm functions to dedicated CacheNameSpaceTrait.

We intend to use the same namespacing algorithm for HPOS orders cache like we have been using before for posts. Moving this structure to a common trait allow us to use the same logic for both posts and HPOS.

vedanshujain committed:
Allow grouping support in cache engines.

This will allow us to add namespacing support directly in the cache engine.

vedanshujain committed:
Remove namespacing from object cache since its supported by cache engine directly.

vedanshujain committed:
Use HPOS feature flag for order cache as well.

vedanshujain committed:
Apply renaming of WPCacheEngine class.

vedanshujain committed:
Fix unit tests for object cache to accomodate recent changes.

vedanshujain committed:
Applied coding standards.

vedanshujain committed:
Applied coding standards.

vedanshujain committed:
Merge fixes.

vedanshujain committed:
Coding standard fix.

vedanshujain committed:
Remove infinite loop by unlinking cache and datastore.

vedanshujain committed:
Clear cache after deleting post.

vedanshujain committed:
Temporary disable cache before backfilling.

vedanshujain committed:
Coding standard fixes.

Konamiman commented on plugins/woocommerce/src/Caches/OrderCache.php:
```diff
@@ -0,0 +1,56 @@
+	 *
+	 * @return array|object|null The retrieved object, or null if it's not possible to retrieve an object by the given id.
+	 */
+	protected function get_from_datastore( $id ) {
+		return null;
+	}
```
This method can go away since it already exists in the base class `ObjectCache` where it's also simply returning `null`.

vedanshujain replied:
Addressed in https://github.com/woocommerce/woocommerce/pull/35014/commits/c678ce72486efecd0ca88bbf909c374e0284fe44

Konamiman commented on plugins/woocommerce/src/Caching/ObjectCache.php:
```diff
@@ -286,41 +258,16 @@ public function get( $id, int $expiration = self::DEFAULT_EXPIRATION, callable $
-		 * @param ObjectCache $cache_instance The instance of ObjectCache whose 'flush` method has been called.
-		 * @param CacheEngine $engine The cache engine in use.
-		 */
-		do_action( "woocommerce_after_flushing_{$this->object_type}_cache", $this, $this->get_cache_engine() );
```
Has this action been removed on purpose?

vedanshujain replied:
Yeah, so now that we are using the same namespacing, this action is no longer reliable. The cache could be flushed by legacy code without it actually hitting here (and also by memcache server restart and other non capturable events). It's much more reliable to hook to order modify methods to capture cache invalidations.

Konamiman commented:
In the first set of testing commands `deep_assoc_array_diff` isn't returning an empty array for me. Not sure if relevant, but the metadata for `$order` and `$r_order` is the same but with different ids:

`$order`:

```
 ["meta_data":protected]=>
  array(2) {
    [0]=>
    object(WC_Meta_Data)#5647 (2) {
      ["current_data":protected]=>
      ...
    [1]=>
    object(WC_Meta_Data)#5648 (2) {
      ["current_data":protected]=>
      ...
```

`$r_order`: (same actual data)

```
 ["meta_data":protected]=>
  array(2) {
    [0]=>
    object(WC_Meta_Data)#5644 (2) {
      ["current_data":protected]=>
      ...
    [1]=>
    object(WC_Meta_Data)#5642 (2) {
      ["current_data":protected]=>
      ...
```

barryhughes commented on plugins/woocommerce/src/Caching/WPCacheEngine.php:
```diff
@@ -0,0 +1,76 @@
+	 *
+	 * @return array|object|null The cached object, or null if there's no object cached under the passed key.
+	 */
+	public function get_cached_object( string $key, string $group = '' ) {
```
Is it possible there is user code out there (perhaps within some sites running on WP Engine) that extends this class/is it safe to change the method in this way (ditto some of the other methods where we change the method signature)?

vedanshujain replied:
Just linking that it was answered here.

barryhughes commented on plugins/woocommerce/src/Caches/OrderCache.php:
```diff
@@ -0,0 +1,56 @@
+	 */
+	protected function validate( $object ): ?array {
+		if ( ! $object instanceof \WC_Abstract_Legacy_Order ) {
+			return array( 'The supplied order is not an instance of WC_Order, ' . gettype( $object ) );
```
```suggestion
			return array( 'The supplied order is not an instance of WC_Abstract_Legacy_Order, ' . gettype( $object ) );
```

Maybe more accurate?

vedanshujain replied:
Addressed in https://github.com/woocommerce/woocommerce/pull/35014/commits/b5ae765b9311a88d94b24154358fbdde59e872d6

vedanshujain committed:
Remove get_from_datastore as it was not being used.

vedanshujain committed:
Use correct class name.

barryhughes reviewed (approved):
Nice work—tested with the revised and updated instructions, all working well 👍🏼

vedanshujain committed:
Merge branch 'trunk' into revert-34992-revert-34396-add/order_cache

Konamiman committed:
Merge branch 'trunk' into revert-34992-revert-34396-add/order_cache
